hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/cache/PublishCacheManager.java

```diff
@@ -99,26 +100,35 @@ public void init() {
 
             definitions.registerListener(events -> {
                 log.debugf("Receiving %s", events);
-                events.stream().forEach(e -> {
-                    log.debugf("Received %s", e);
-                    String tenantId = e.getTargetTenantId();
-                    String triggerId = e.getTargetId();
-                    TriggerKey triggerKey = new TriggerKey(tenantId, triggerId);
-                    publishCache.startBatch();
-                    publishDataIdsCache.startBatch();
-                    Set<String> oldDataIds = publishDataIdsCache.get(triggerKey);
-                    removePublishCache(tenantId, triggerId, oldDataIds);
-                    if (e.getType().equals(TRIGGER_CONDITION_CHANGE)) {
-                        Set<String> newDataIds = e.getDataIds();
-                        publishDataIdsCache.put(triggerKey, newDataIds);
-                        addPublishCache(tenantId, triggerId, newDataIds);
-                    }
-                    publishDataIdsCache.endBatch(true);
-                    publishCache.endBatch(true);
-                });
+                events.stream()
+                        .sorted((e1, e2) -> e1.getType() == TRIGGER_REMOVE ? 1 : -1) // process removals last
```

Comment:
I think I see where this is coming.
Unless I miss something I see risky to order the events and process the TRIGGER_REMOVE at last.
I have the feeling it could break things.

For example, in the code we might have:
[1] Delete trigger A
[2] Create new trigger A
[3] Update conditions on trigger A

In previous commit the events type were moved to a list to preserve the order of processing.
With the current code I guess that [1] can be executed at last and then break the order of notifications.

Another thing that it can impact, is that in the tests we don't count that a modification needs time in the engine to process the changes. Sometimes we do the [1][2][3] steps in a test plus [4] send data, and results are not expected because async processing didn't finish.

We can comment off line, perhaps if you have an example we can take a look, I can check if I am understanding the issue correctly.